### PR TITLE
Document versioning expectations since v1.0

### PR DESCRIPTION
**Pull request by josevalim**

### CODE REVIEW

File: lib/elixir/pages/Compatibility and Deprecations.md [hunk 1]
```diff
@@ -1,5 +1,25 @@
...
+
+  * Bugs: if an application has undesired behaviour, a program that depends on the buggy behaviour may break if the bug is fixed. We reserve the right to fix such bugs.
+
+  * Compiler front-end: improvements may be done to the compiler, introducing new warnings for ambiguous modes and providing more detailed error messages. Those can lead to compilation errors (when running with `--warning-as-errors`) or tooling failures when asserting on specific error messages (although one should avoid such). We reserve the right to do such improvements.
+
+  * Imports: new functions may be added to the `Kernel` module, which is auto-imported. They may collide with local functions defined in your modules. Collisions can be resolved in a backwards compatible fashion using `import Kernel, except: [...]` with a list of all functions you don't want imported from `Kernel`. We reserve the right to do such additions.
+
+In order to continue evolving the language without introducing breaking changes, Elixir will rely on deprecations to demote certain practices and promote new ones. Our deprecation policy is outlined in the "Deprecations" section.
```
Comment (michalmuskala) [inline lib/elixir/pages/Compatibility and Deprecations.md]: It would be nice to make this a link

File: lib/elixir/pages/Compatibility and Deprecations.md [hunk 2]
```diff
@@ -1,5 +1,25 @@
...
+
+Elixir binaries are guaranteed to be backwards compatible within the same branch. Code compiled with v1.0 shall work with Elixir v1.0.1 runtime but code compiled with Elixir v1.0 should be recompiled to work with Elixir v1.1.0 onwards. The Mix build tool automatically tracks those changes.
```
Comment (michalmuskala) [inline lib/elixir/pages/Compatibility and Deprecations.md]: Is that actually true, though? This completely ignores the factor of Erlang version, which I think is actually more important here.

File: lib/elixir/pages/Compatibility and Deprecations.md [hunk 3]
```diff
@@ -1,5 +1,25 @@
...
+
+The only exception to the rules above are experimental features, which will be explicitly marked as such, and do not provide any compatibility guarantee until they are stabilized.
```
Comment (michalmuskala) [inline lib/elixir/pages/Compatibility and Deprecations.md]: I think this is a very good point. Do we have a standard way to mark features as experimental?

File: lib/elixir/pages/Compatibility and Deprecations.md (hunk 2 above)
Comment (josevalim) [reply lib/elixir/pages/Compatibility and Deprecations.md]: We do consider the erlang version since v1.6.

File: lib/elixir/pages/Compatibility and Deprecations.md (hunk 3 above)
Comment (josevalim) [reply lib/elixir/pages/Compatibility and Deprecations.md]: No because we never used but it would probably be a huge disclaimer in the documentation.

File: lib/elixir/pages/Compatibility and Deprecations.md [hunk 4]
```diff
@@ -1,5 +1,25 @@
...
 
+## Compatibility between Elixir v1.x versions
+
+Elixir v1.x releases are backwards compatible: well-defined behaviours and documented APIs in a given release will continue working on future releases.
+
+Although we expect the vast majority of programs to remain compatible over time, it is impossible to guarantee that no future change will break any program. Under some unlikely circumstances, we may introduce changes that break existing code:
+
+  * Security: a security issue in the implementation may arise whose resolution requires backwards incompatible changes. We reserve the right to address such security issues.
+
+  * Bugs: if an application has undesired behaviour, a program that depends on the buggy behaviour may break if the bug is fixed. We reserve the right to fix such bugs.
```
Comment (whatyouhide) [inline lib/elixir/pages/Compatibility and Deprecations.md]: > if an application has undesired behaviour

This is a bit unclear to me. Do you mean an Elixir application (logger, mix) when you say that? Maybe we can rephrase this as:

> Bugs: if your application relies on a bug in Elixir, then your application might break if the bug is fixed. We reserve the right to fix such bugs.

**Comment by josevalim** [comment]
All concerns addressed /cc @michalmuskala and @whatyouhide

File: lib/elixir/pages/Compatibility and Deprecations.md [hunk 5]
```diff
@@ -1,5 +1,25 @@
...
+
+  * Security: a security issue in the implementation may arise whose resolution requires backwards incompatible changes. We reserve the right to address such security issues.
+
+  * Bugs: if an API has undesired behaviour, a program that depends on the buggy behaviour may break if the bug is fixed. We reserve the right to fix such bugs.
+
+  * Compiler front-end: improvements may be done to the compiler, introducing new warnings for ambiguous modes and providing more detailed error messages. Those can lead to compilation errors (when running with `--warning-as-errors`) or tooling failures when asserting on specific error messages (although one should avoid such). We reserve the right to do such improvements.
+
+  * Imports: new functions may be added to the `Kernel` module, which is auto-imported. They may collide with local functions defined in your modules. Collisions can be resolved in a backwards compatible fashion using `import Kernel, except: [...]` with a list of all functions you don't want imported from `Kernel`. We reserve the right to do such additions.
```
Comment (eksperimental) [inline lib/elixir/pages/Compatibility and Deprecations.md]: you don't want TO BE imported

File: lib/elixir/pages/Compatibility and Deprecations.md [hunk 6]
```diff
@@ -1,5 +1,25 @@
...
+
+In order to continue evolving the language without introducing breaking changes, Elixir will rely on deprecations to demote certain practices and promote new ones. Our deprecation policy is outlined [in the "Deprecations" section](#deprecations).
```
Comment (eksperimental) [inline lib/elixir/pages/Compatibility and Deprecations.md]: I think "in the" should not be part of the link

File: lib/elixir/pages/Compatibility and Deprecations.md [hunk 7]
```diff
@@ -1,5 +1,25 @@
...
+
+Elixir binaries are guaranteed to be backwards compatible within the same branch. Code compiled with v1.0 shall work with Elixir v1.0.1 runtime but code compiled with Elixir v1.0 should be recompiled to work with Elixir v1.1.0 onwards. The Mix build tool automatically tracks those changes.
```
Comment (eksperimental) [inline lib/elixir/pages/Compatibility and Deprecations.md]: FROM v1.1.0 onwards

**Comment by josevalim** [comment]
Thanks @eksperimental! :heart:

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by eksperimental** [comment]
No worries
